### PR TITLE
[Kernel] add l2norm triton kernel

### PR DESCRIPTION
**Pull request by OsirisDuan**
### What this PR does / why we need it?
This pull request introduces an L2 normalization kernel implemented in Triton, specifically optimized for Ascend NPUs. 
### Does this PR introduce _any_ user-facing change?
No, this PR does not introduce any user-facing changes.
### How was this patch tested?

- vLLM version: v0.13.0
- vLLM main: https://github.com/vllm-project/vllm/commit/bc0a5a0c089844b17cb93f3294348f411e523586

### CODE REVIEW

File: vllm_ascend/ops/triton/fla/l2norm.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+):
+    i_t = tl.program_id(0)
+    x += i_t * D
+    y += i_t * D
+    # Compute mean and variance
+    cols = tl.arange(0, BD)
+    mask = cols < D
+    b_x = tl.load(x + cols, mask=mask, other=0.0).to(tl.float32)
+    b_var = tl.sum(b_x * b_x, axis=0)
+    b_rstd = 1 / tl.sqrt(b_var + eps)
```
[the remark this answers is not in the document]
Comment (OsirisDuan) [reply vllm_ascend/ops/triton/fla/l2norm.py]: In fact, in the QWen3-Next model, we only use the kernel `l2norm_fwd_kernel2_loop`, so the other kernels remain consistent with those in VLLM.

File: vllm_ascend/ops/triton/fla/l2norm.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    T,
+    D: tl.constexpr,
+    BT: tl.constexpr,
+    BD: tl.constexpr,
+):
+    i_t = tl.program_id(0)
+    p_x = tl.make_block_ptr(x, (T, D), (D, 1), (i_t * BT, 0), (BT, BD), (1, 0))
+    b_x = tl.load(p_x, boundary_check=(0, 1)).to(tl.float32)
+    b_var = tl.sum(b_x * b_x, axis=1)
+    b_y = b_x / tl.sqrt(b_var + eps)[:, None]
```
[the remark this answers is not in the document]
Comment (OsirisDuan) [reply vllm_ascend/ops/triton/fla/l2norm.py]: In fact, in the QWen3-Next model, we only use the kernel `l2norm_fwd_kernel2_loop`, so the other kernels remain consistent with those in VLLM.

File: vllm_ascend/ops/triton/fla/l2norm.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    base_row = tl.program_id(0) * (NUM_CHUNKS * MBLOCK)
+    rindex = tl.arange(0, N)[None, :]
+
+    for chunk in range(NUM_CHUNKS):
+        row_idx = base_row + chunk * MBLOCK + tl.arange(0, MBLOCK)[:, None]
+        xmask = row_idx < M
+
+        xs = tl.load(X + (rindex + N * row_idx), xmask).to(tl.float32)
+        square = xs * xs
+        square_sum = tl.sum(tl.where(xmask, square, 0), 1)[:, None]
```
[the remark this answers is not in the document]
Comment (OsirisDuan) [reply vllm_ascend/ops/triton/fla/l2norm.py]: Accepted

File: vllm_ascend/ops/triton/fla/l2norm.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+USE_DEFAULT_FLA_NORM = False
+
+@triton.autotune(
+    configs=[
+        triton.Config({}, num_warps=num_warps) for num_warps in [1, 2, 4, 8, 16, 32]
+    ],
+    key=["D"],
+)
+@triton.jit
+def l2norm_fwd_kernel1(
```
Comment (momo609) [inline vllm_ascend/ops/triton/fla/l2norm.py]: add op test.

File: vllm_ascend/ops/triton/fla/l2norm.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+    T, D = x.shape[0], x.shape[-1]
+    # rstd = torch.empty((T,), dtype=torch.float32, device=x.device)
+    # Less than 64KB per feature: enqueue fused kernel
+    MAX_FUSED_SIZE = 65536 // x.element_size()
+    BD = min(MAX_FUSED_SIZE, triton.next_power_of_2(D))
+    if D > BD:
+        raise RuntimeError("This layer doesn't support feature dim >= 64KB.")
+
+    if not USE_DEFAULT_FLA_NORM:
+        MBLOCK = 69
```
Comment (weijinqian0) [inline vllm_ascend/ops/triton/fla/l2norm.py]: Can this branch be deleted?

**Comment by weijinqian0** [comment]
need golden test

File: vllm_ascend/ops/triton/fla/l2norm.py (hunk 5 above)
Comment (OsirisDuan) [reply vllm_ascend/ops/triton/fla/l2norm.py]: done

File: vllm_ascend/ops/triton/fla/l2norm.py (hunk 4 above)
Comment (OsirisDuan) [reply vllm_ascend/ops/triton/fla/l2norm.py]: The kernel was deleted because it was not in use; the kernels that are in use will have test cases added later.

**Review by Yikun — CHANGES_REQUESTED** [review]
as @weijinqian0 required, pls add e2e test

also cc @MengqingCao

**Comment by OsirisDuan** [comment]
> as @weijinqian0 required, pls add e2e test
> 
> also cc @MengqingCao

done

**Review by MengqingCao — APPROVED** [review]

File: .github/workflows/_e2e_test.yaml [hunk 6]
```diff
@@ -103,6 +103,7 @@ jobs:
           # We found that if running aclgraph tests in batch, it will cause AclmdlRICaptureBegin error. So we run
           # the test separately.
 
+          pytest -sv --durations=0 tests/e2e/nightly/ops/triton/test_l2norm.py
```
Comment (Yikun) [inline .github/workflows/_e2e_test.yaml]: We could consider run all ops in followup PR.

```
pytest -sv --durations=0 tests/e2e/nightly/ops/triton/
```

cc @MengqingCao

**Review by Yikun — APPROVED** [review]
```
tests/e2e/nightly/ops/triton/test_l2norm.py::test_l2norm[B1-T63-H1-D60-torch.float32] PASSED
tests/e2e/nightly/ops/triton/test_l2norm.py::test_l2norm[B2-T500-H4-D64-torch.float32] PASSED
tests/e2e/nightly/ops/triton/test_l2norm.py::test_l2norm[B2-T1000-H2-D100-torch.float32] PASSED
tests/e2e/nightly/ops/triton/test_l2norm.py::test_l2norm[B3-T1024-H4-D128-torch.float32] PASSED
```

Thank you for your patience and for tolerating my -1. Your PR provides a good example for contributions to other triton operators.

Thank you for your contribution.

File: .github/workflows/_e2e_test.yaml (hunk 6 above)
Comment (OsirisDuan) [reply .github/workflows/_e2e_test.yaml]: done

**Review by MengqingCao — APPROVED** [review]

File: vllm_ascend/ops/triton/fla/l2norm.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+# Copyright (c) 2023-2025, Songlin Yang, Yu Zhang
+
+import torch
+from vllm.triton_utils import tl, triton
+
+from vllm_ascend.ops.triton.triton_utils import get_vectorcore_num
+
+
+@triton.jit
+def l2norm_fwd_kernel2_loop(X, Y, eps, M, N: tl.constexpr,
```
Comment (Yikun) [inline vllm_ascend/ops/triton/fla/l2norm.py:18]: I just found this ops not been used yet, Pls submit a new PR remove this, when re adding new ops, ops/test/workflow should be added together.

Comment (OsirisDuan) [reply vllm_ascend/ops/triton/fla/l2norm.py]: We used this kernel at line 60.
